### ISSUE TITLE
cacert configuration is not working

### ISSUE THREAD

**Issue by Seb0042** [vmware-tanzu/velero#2589]
**What steps did you take and what happened:**
I installed velero with the --cacert configuration, like this:
velero install --use-restic --no-default-backup-location --use-volume-snapshots=false --secret-file miniocred.yaml --namespace=myns --cacert=ca.pem

In addition, I created a backup location in wich I add a caCert value with the cert chain (via kubectl edit backupstoragelocations.velero.io default).

**What did you expect to happen:**
No error in velero pod about x509 connection

**The output of the following commands will help us better understand what's going on**:

time="2020-05-29T09:06:13Z" level=error msg="Error checking repository for stale locks" controller=restic-repository error="error running command=restic unlock --repo=s3:https://minio.mydomain:9000/velero/restic/kube-system --password-file=/tmp/velero-restic-credentials-kube-system573266569 --cache-dir=/scratch/.cache/restic, stdout=, stderr=Fatal: unable to open config file: Stat: Get https://minio.mydomain:9000/velero/?location=: x509: certificate signed by unknown authority\nIs there a repository at the following location?\ns3:https://minio.mydomain:9000/velero/restic/kube-system\n: exit status 1" error.file="/github.com/vmware-tanzu/velero/pkg/restic/repository_manager.go:293" error.function="github.com/vmware-tanzu/velero/pkg/restic.(*repositoryManager).exec" logSource="pkg/controller/restic_repository_controller.go:141" name=kube-system-default-xq46p namespace=production-system

**Anything else you would like to add:**
The only way I made it work was to modify the velero deployment and add the bundle in /etc/ssl/certs. But this doesn't propagate to restics pods.

**Environment:**
- Velero version (use `velero version`):  v1.4.0
- Velero features (use `velero client config get features`): features: <NOT SET>
- Kubernetes version (use `kubectl version`): "v1.18.2"
- Kubernetes installer & version:  kubeadm v1.18.2
- Cloud provider or hardware configuration: VMware virtual machines
- OS (e.g. from `/etc/os-release`): CentOS 7.8.2003

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by Seb0042** [comment]
```
apiVersion: velero.io/v1
kind: BackupStorageLocation
metadata:
  name: minio
  namespace: myns
spec:
  accessMode: ReadWrite
  config:
    caCert: |
      -----BEGIN CERTIFICATE-----
      MIIEsTCCA5mgAwIBAgIQCKWiRs1LXIyD1wK0u6tTSTANBgkqhkiG9w0BAQsFADBh
      xxxxxxxxxxxxxxxxxxxxxxxx
      -----END CERTIFICATE-----
    region: us-east-1
    s3ForcePathStyle: "true"
    s3Url: https://minio.mydomain:9000
  objectStorage:
    bucket: velero
  provider: aws
```

**Comment by ashish-amarnath** [comment]
The `caCert` field has to be at the path `spec.objectStorage.caCert` 
https://github.com/vmware-tanzu/velero/blob/master/pkg/apis/velero/v1/backup_storage_location.go#L70
This should not be base64 encoded.
`caCert` is of type `[]byte`. Can you please tell us what you mean by
> If I try to move it "as it is" I can't save it because it suppose to be a type "byte".

Is there an error that you see?

**Comment by Seb0042** [comment]
Yes, when I edit the backupstoragelocation with kubectl.
If I do like this, it won't work. I will have the following error:
# * spec.objectStorage.caCert: Invalid value:  (certificate is written here)
spec.objectStorage.caCert in body must be of type byte 
```
$ kubectl edit backupstoragelocations.velero.io minio 
```
```
spec:
  accessMode: ReadWrite
  config:
    region: us-east-1
    s3ForcePathStyle: "true"
    s3Url: https://minio.mydomain:9000
objectStorage:
    bucket: velero
    caCert: |
      -----BEGIN CERTIFICATE-----
      xxxxx
      -----END CERTIFICATE-----
  provider: aws
```

**Comment by ashish-amarnath** [comment]
@Seb0042 Sorry about giving you incorrect info earlier.
You were right in base64 encoding the cert. 
The yaml for your BSL would look like

```
apiVersion: velero.io/v1
kind: BackupStorageLocation
metadata:
  name: minio
  namespace: myns
spec:
  accessMode: ReadWrite
  config:
    region: us-east-1
    s3ForcePathStyle: "true"
    s3Url: https://minio.mydomain:9000
  objectStorage:
    bucket: velero
    caCert: "BASE64 ENCODED CERT VALUE IN ONE LINE"
  provider: aws
```

This should resolve your issue

**Comment by Seb0042** [comment]
It doesn't, I have the error message that I wrote above (pem cert can't be
loaded).

Le mer. 3 juin 2020 à 01:20, Ashish Amarnath <notifications@github.com> a
écrit :

**Comment by Seb0042** [comment]
OK, did few tests, maybe it was a typo or something, now I can have caCert both in config and objectStorage without error.
But the issue is still the same, if I don't put my ca file in /etc/ssl/certs I will have an error about wrong certificate.

**Comment by Seb0042** [comment]
OK, I did more tests and investigations.

So, because of name resolution and reaching port 9000 of minio, I was launching my velero command from inside the velero pod. The x509 error was comming from that. It's velero as a client which was failing.

So, now, it seems to work if I do the following:
- add caCert to config.
- add caCert to storageObject.
- use the velero command from a node of my cluster, with name resolution "minio.mydomain" as the IP of the minio service.

In the past, first, I was using the velero command from my admin linux server. I had to use the external IP for minio, but in this case it wouldn't work (port 9000 not open from external, only 443 via traefik). 
It's why next I use the velero command from the velero pod, and therefore the issue with the x509 certificate as a client.

So it leaves me with a choice:
- use it from my admin server but I have to open the 9000 port on external , or change the native port of minio to 443.
- have the velero command installed on my kubernetes nodes to be able to reach the minio on port 9000.

Any suggestion ? I will close the issue then because it was more related to client configuration.
